source/agora/node/TransactionRelayer.d

```diff
@@ -0,0 +1,462 @@
+
+    private Transaction[] getRelayTransactions () @safe nothrow
+    {
+        immutable now = this.clock.networkTime();
+        Transaction[] txs_gathered;
+        reserve(txs_gathered, this.config.node.relay_tx_max_num);
+
+        while (!this.txs.empty() &&
+                ((this.config.node.relay_tx_max_num == 0) ||
+                 (txs_gathered.length < this.config.node.relay_tx_max_num)))
+        {
+            TxHolder tx_holder = this.txs.front();
+            if (tx_holder.expiry >= now)
+            {
+                Transaction tx = this.pool.getTransactionByHash(tx_holder.hash);
+                if (tx != Transaction.init)
+                    txs_gathered ~= tx;
+            }
+
+            this.tx_hashes.remove(tx_holder.hash);
+            this.txs.removeFront();
+        }
+        return txs_gathered;
+    }
+
+    /// Sends an array of transactions ordered by fee to known network clients.
+    public void relayTransactions () @safe nothrow
+    {
+        this.getRelayTransactions().each!(tx => (*this.clients).each!(node_info => node_info.client.sendTransaction(tx)));
+    }
+
+    /// Cleans expired entries from the internal datastructures.
+    private void cleanRelayQueue () @safe
+    {
+        immutable now = this.clock.networkTime();
+
+        // Cannot iterate through a RedBlackTree, and delete its elements at the
+        // same time, so this is done in 2 steps.
+        auto tx_holders = this.txs[].filter!(tx_holder => tx_holder.expiry < now).array();
+        tx_holders.each!((tx_holder)
+        {
+            this.tx_hashes.remove(tx_holder.hash);
+            this.txs.removeKey(tx_holder);
+        });
+    }
+
+    /***************************************************************************
+
+        Adds a transaction to the relay queue.
+
+        Params:
+            tx = transaction to be added to the relay queue
+
+        Returns:
+            null, if there was no error, otherwise string detailing the error
+
+    ***************************************************************************/
+
+    private string addTransactionImp (in Transaction tx) @safe nothrow
+    {
+        auto tx_hash = tx.hashFull();
+        if (tx_hash in this.tx_hashes)
+            return TransactionRelayError.TxAlreadyAdded;
+
+        Amount fee;
+        if (auto fee_err = this.getAdjustedTXFee(tx, fee))
+            return fee_err;
+        else if (!fee.isValid())
+            return TransactionRelayError.TxFeeInvalid;
+
+        if (fee < this.config.node.relay_tx_min_fee)
+            return TransactionRelayError.TxFeeTooLow;
+
+        this.txs.insert(TxHolder(fee, tx_hash,
+            this.clock.networkTime() + this.config.node.relay_tx_cache_exp.total!"seconds"));
+        this.tx_hashes.put(tx_hash);
+
+        return null;
+    }
+
+    /***************************************************************************
+
+        Adds a transaction to the relay queue. Unlike addTransactionImp, this
```

Comment:
```suggestion
        Adds a transaction to the relay queue.
        
        Unlike addTransactionImp, this [...]
```

I would phrase it as "can be overridden by child class to perform additional behavior", or the likes, since I suppose you did this to override it in UT ?